Refactor and fix xilinx_zynqmp SoC definition

stephanosio opened the pull request:
```
ZynqMP SoC embeds two separate processor types: Cortex-R for RPU and
Cortex-A for APU.

Since the current Zephyr architecture cannot support AMP of Cortex-R
and Cortex-A within one project, the RPU and APU should be considered
separate platforms.

This commit relocates the device tree nodes that are not common between
RPU and APU to a separate dtsi file (zynqmp_rpu.dtsi).

When Cortex-A53 APU support is added in the future, an additional dtsi
file (zynqmp_apu.dtsi) for specifying the APU device tree should be
added.
```

```
This commit fixes the following problems with the RPU device tree:

1. The core type of the RPU of ZynqMP SoC is Cortex-R5F, not
  Cortex-R4.

2. RPU and APU use different interrupt controllers (PL390 GICv1 and
  GIC-400 GICv2, respectively) mapped to the same CPU local bus address
  region but with different offsets for the distributor and CPU
  interrupt control register sets. The GIC address mapping specified by
  the current dts is that of an APU and does not apply to the PL390
  GICv1 of an RPU (refer to the "Zynq UltraScale+ Devices Register
  Reference" document from Xilinx for more information).
```

```
The Xilinx ZynqMP SoC embeds both Cortex-R "RPU" and Cortex-A "APU"
cores.

Since the current Zephyr architecture cannot support AMP of Cortex-R
and Cortex-A within the same project, the RPU and APU should be
considered separate platforms and handled accordingly.

This commit re-purposes the SOC_XILINX_ZYNQMP symbol as a helper symbol
indicating that Xilinx ZynqMP SoC is used, and adds a new symbol,
SOC_XILINX_ZYNQMP_RPU, for specifying the actual build target platform.

When Cortex-A support is added in the future, SOC_XILINX_ZYNQMP_APU
symbol should be added and used to conditionally handle APU-specific
code.
```

This PR is in preparation for the issue #20217.

stephanosio committed:
dts: xilinx_zynqmp: Refactor dts to specify RPU and APU separately.

ZynqMP SoC embeds two separate processor types: Cortex-R for RPU and
Cortex-A for APU.

Since the current Zephyr architecture cannot support AMP of Cortex-R
and Cortex-A within one project, the RPU and APU should be considered
separate platforms.

This commit relocates the device tree nodes that are not common between
RPU and APU to a separate dtsi file (zynqmp_rpu.dtsi).

When Cortex-A53 APU support is added in the future, an additional dtsi
file (zynqmp_apu.dtsi) for specifying the APU device tree should be
added.

For more details, refer to the issue #20217.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
dts: xilinx_zynqmp: Fix incorrect RPU device tree specification.

This commit fixes the following problems with the RPU device tree:

1. The core type of the RPU of ZynqMP SoC is Cortex-R5F, not
  Cortex-R4.

2. RPU and APU use different interrupt controllers (PL390 GICv1 and
  GIC-400 GICv2, respectively) mapped to the same CPU local bus address
  region but with different offsets for the distributor and CPU
  interrupt control register sets. The GIC address mapping specified by
  the current dts is that of an APU and does not apply to the PL390
  GICv1 of an RPU (refer to the "Zynq UltraScale+ Devices Register
  Reference" document from Xilinx for more information).

For more details, refer to the issue #20217.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
soc: arm: xilinx_zynqmp: Refactor for multi-arch support.

The Xilinx ZynqMP SoC embeds both Cortex-R "RPU" and Cortex-A "APU"
cores.

Since the current Zephyr architecture cannot support AMP of Cortex-R
and Cortex-A within the same project, the RPU and APU should be
considered separate platforms and handled accordingly.

This commit re-purposes the SOC_XILINX_ZYNQMP symbol as a helper symbol
indicating that Xilinx ZynqMP SoC is used, and adds a new symbol,
SOC_XILINX_ZYNQMP_RPU, for specifying the actual build target platform.

When Cortex-A support is added in the future, SOC_XILINX_ZYNQMP_APU
symbol should be added and used to conditionally handle APU-specific
code.

For more details, refer to the issue #20217.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio commented:
cc @bbolen @carlocaione

stephanosio commented:
Here is a quick note on how to test this in qemu_cortex_r5 (in context of #20217):

1. Download and install the test SDK release based on https://github.com/zephyrproject-rtos/sdk-ng/pull/138 from the following link:
    https://github.com/stephanosio/zephyr-sdk-ng/releases/tag/stephanos-v0.11.0-alpha-arm_generic_fdt

2. Set QEMU_BIN_PATH to the test SDK path (this allows `ninja run` to use the `qemu-system-aarch64` from the test SDK release).
    ```
    export QEMU_BIN_PATH=/opt/zephyr-sdk/2.2/sysroots/x86_64-pokysdk-linux/usr/bin
    ```

2. Check out this PR branch (stephanosio:fix_zynqmp).

stephanosio commented:
Here is a quick note on how to test this in qemu_cortex_r5 (in context of #20217):

1. Download and install the test SDK release based on https://github.com/zephyrproject-rtos/sdk-ng/pull/138 from the following link:
    https://github.com/stephanosio/zephyr-sdk-ng/releases/tag/stephanos-v0.11.0-alpha-arm_generic_fdt

2. Set QEMU_BIN_PATH to the test SDK path (this allows `ninja run` to use the `qemu-system-aarch64` from the test SDK release).
    ```
    export QEMU_BIN_PATH=/opt/zephyr-sdk/2.2/sysroots/x86_64-pokysdk-linux/usr/bin
    ```

3. Check out this PR.
    ```
    git fetch upstream pull/20267/head:test_fix_zynqmp
    git checkout test_fix_zynqmp
    ```

4. Rebase onto PR #20298 (use `arm-generic-fdt` QEMU machine type for qemu_cortex_r5 emulation).
    ````
    git fetch upstream pull/20298/head:qemu_cortex_r5_use_xilinx_qemu
    git rebase qemu_cortex_r5_use_xilinx_qemu
    ````

5. Build and run any of the kernel tests (e.g. tests/kernel/threads/thread_apis).

With this PR, most kernel tests should pass (some may fail due to the bugs in the Cortex-R port that are still being addressed in the #19698; without this PR, the kernel tests will be stuck).

On a side note, the reason that this currently works without using the correct GIC driver is that the GIC-400 GICv2 driver currently being used is mostly compatible with the PL390 GICv1 (the current driver sets a few extra registers that do not exist on GICv1, but this does not really hinder GIC operation).

stephanosio commented:
Force pushing to resolve merge conflicts.

carlocaione commented on boards/arm/qemu_cortex_r5/qemu_cortex_r5.dts:
```diff
@@ -6,7 +6,7 @@
+#include <arm/xilinx/zynqmp_rpu.dtsi>
 
 / {
 	model = "QEMU Cortex-R5";
```
Do we want to specify RPU in the model and compatible as well?

stephanosio replied:
For `model`, this is a "QEMU Cortex-R5" board that just happens to be using Xilinx ZynqMP, so I don't think "RPU" (Xilinx ZynqMP-specific term) belongs there. In addition, "Cortex-R5" pretty much implies "RPU." This also aligns with what `qemu_cortex_m0` and `qemu_cortex_m3` are doing.

As for `compatible`, this field has no real function (at least for now) and the common trend seems to be to specify the compatible SoC name. Since the SoC is `xlnx,zynqmp` regardless of RPU or APU, I think we should leave it as is (`#include dtsi` is what really determines which DT is used by board). Regarding the `-qemu` suffix, this probably should be removed unless we have a specific reason to differentiate qemu vs. non-qemu for zynqmp (e.g. as was done in as `qemu_cortex_m3`).

stephanosio committed:
boards: qemu_cortex_r5: Remove ignore tags for working tests

This commit removes the ignore tags for the tests that work after the
changes in the PR #20267.

In the future, this ignored testing tag list will be further reduced
as critical bugs for the qemu_cortex_r5 platform are addressed
(see #20217).

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio commented:
CI tests are run for `qemu_cortex_r5` now. Please review and merge this.

stephanosio commented:
Adding "bug" label, because this PR is fixing a fundamental design flaw as well as incorrect interrupt controller specification in the device tree that causes multithreading to not work on `xilinx_zynqmp`.

stephanosio commented:
Force pushing to resolve merge conflicts.

stephanosio commented:
@carlocaione @ioannisg Could you review this PR? This is currently blocking all the other Cortex-R PRs as well as the ZynqMP support in the AArch64 PR.

carlocaione commented:
LGTM
